### PR TITLE
Feature: Add libgeos in third party, verify existence of low emission zone (LEZ) on path using mode Car and CarNopark  

### PR DESCRIPTION
**Pull request by xlqian**
* Use third_party libgeos
* Load one or more LEZ files by using parameter ASGARD_LEZ_SHAPES_PATH=.. while executing asgard
* For a path in a street network using modes Car as well as CarNopark, verify existence of LEZ.
* Display the node low_emission_zone in the street network section as well as journey with LEZ as below:
`"low_emission_zone": {
     "on_path": true
}
` 
For more details: https://navitia.atlassian.net/browse/NAV-1787

PR navitia to display the node low_emission_zone: https://github.com/hove-io/navitia/pull/3959

### CODE REVIEW

File: asgard/direct_path_response_builder.cpp [hunk 1]
```diff
@@ -257,6 +264,24 @@ void build_mono_modal_journey(pbnavitia::Journey& journey,
     section->set_end_date_time(departure_posix_time + section_duration);
 
     compute_geojson(shape, *section);
+
+    // Verify existence of low emission zone for the path and add node low_emission_zone for Car and CarNoPark
+    if ((lez_list.size() > 0) &&
+        (section->street_network().mode() == pbnavitia::StreetNetworkMode::Car ||
```
Comment (pbench) [inline asgard/direct_path_response_builder.cpp:270]: for mode `Car` does the `shape` includes the walking part between the parking and the stop point ?

Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp:270]: Non, it includes only the driving path

File: asgard/direct_path_response_builder.cpp [hunk 2]
```diff
@@ -557,6 +584,26 @@ void compute_geojson(const std::vector<midgard::PointLL>& list_geo_points, pbnav
...
+    for (const auto& point : list_geo_points) {
+        cs->add(Coordinate(point.lng(), point.lat()));
+    }
+    auto ls = factory->createLineString(std::move(cs));
+    try {
+        for (size_t idx = 0; idx < lez_list.size(); ++idx) {
+            if (ls->intersects(lez_list[idx].get())) {
+                return true;
+            }
+        }
```
Comment (xlqian) [inline asgard/direct_path_response_builder.cpp:599]:
```suggestion
        return std::any_of(lez_list.cbegin(), lez_list.cend(), [](const auto& lez){ return ls->intersects(lez.get())})
```

**Review by pbench — APPROVED** [review]

File: asgard/handler.cpp [hunk 3]
```diff
@@ -264,6 +277,31 @@ void Handler::log_elapsed_time(const std::chrono::high_resolution_clock::time_po
     LOG_INFO("It took " + std::to_string(elapsed_time.count()) + " ms to handle request: " + request_id);
 }
 
+void Handler::read_lez_shapes() {
+    GeometryFactory::Ptr factory = GeometryFactory::create();
+    GeoJSONReader reader(*factory);
+    std::string dir_name = asgard_lez_shapes_path.get();
+    struct dirent *d;
```
Comment (xlqian) [inline asgard/handler.cpp:284]: dirent is an interface in C, use boost::filesystem for C++ :) https://www.cppstories.com/2019/04/dir-iterate/#boost-filesystem

File: asgard/tests/handler_test.cpp
Comment (woshilapin) [inline asgard/tests/handler_test.cpp:1]: Should we add a test for handling the input file?

File: asgard/tests/direct_path_response_builder_test.cpp
Comment (woshilapin) [inline asgard/tests/direct_path_response_builder_test.cpp:1]: Can we add a test with real LEZ and test that the response does contains `"low_emission_zone":{"on_path":true}`?

File: asgard/direct_path_response_builder.cpp [hunk 4]
```diff
@@ -257,6 +264,24 @@ void build_mono_modal_journey(pbnavitia::Journey& journey,
...
+    if ((lez_list.size() > 0) &&
+        (section->street_network().mode() == pbnavitia::StreetNetworkMode::Car ||
+        section->street_network().mode() == pbnavitia::StreetNetworkMode::CarNoPark)) {
+        auto exists_lez = exists_low_emission_zone(shape, lez_list);
+        auto* j_lez = journey.mutable_low_emission_zone();
+        auto* s_lez = section->mutable_low_emission_zone();
+        if (exists_lez) {
+            s_lez->set_on_path(true);
+            j_lez->set_on_path(true);
+            LOG_INFO("Path intersects LEZ ");
```
[the remark this answers is not in the document]
Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp:278]: :+1:

File: asgard/asgard_conf.h [hunk 5]
```diff
@@ -58,6 +59,7 @@ struct AsgardConf {
         metrics_binding = get_config<std::string>("ASGARD_METRICS_BINDING", std::string("0.0.0.0:8080"));
         valhalla_service_url = get_config<std::string>("ASGARD_VALHALLA_SERVICE_URL", boost::none);
         auto valhalla_conf_json = get_config<std::string>("ASGARD_VALHALLA_CONF", std::string("/data/valhalla/valhalla.json")).get();
+        asgard_lez_shapes_path = get_config<std::string>("ASGARD_LEZ_SHAPES_PATH", boost::none);
```
[the remark this answers is not in the document]
Comment (kadhikari) [reply asgard/asgard_conf.h:62]: :+1:

File: asgard/direct_path_response_builder.cpp [hunk 6]
```diff
@@ -557,6 +584,26 @@ void compute_geojson(const std::vector<midgard::PointLL>& list_geo_points, pbnav
     }
 }
 
+bool exists_low_emission_zone(const std::vector<midgard::PointLL>& list_geo_points, const std::vector<std::unique_ptr<Geometry>>& lez_list) {
+    GeometryFactory::Ptr factory = GeometryFactory::create();
+    auto cs = std::make_unique<CoordinateSequence> ();
```
[the remark this answers is not in the document]
Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp:589]: :+1:

File: asgard/direct_path_response_builder.cpp [hunk 7]
```diff
@@ -557,6 +584,26 @@ void compute_geojson(const std::vector<midgard::PointLL>& list_geo_points, pbnav
...
+    }
+    auto ls = factory->createLineString(std::move(cs));
+    try {
+        for (size_t idx = 0; idx < lez_list.size(); ++idx) {
+            if (ls->intersects(lez_list[idx].get())) {
+                return true;
+            }
+        }
+    } catch (...) {
+        LOG_ERROR("Error occurred while calculating LEZ Geometry");
```
[the remark this answers is not in the document]
Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp:601]: :+1:

File: asgard/direct_path_response_builder.cpp (hunk 2 above)
Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp:599]: Done differently the above suggestion didn't work

File: asgard/direct_path_response_builder.cpp [hunk 8]
```diff
@@ -442,10 +469,10 @@ void build_journey_response(pbnavitia::Response& response,
     case pbnavitia::StreetNetworkMode::Bike:
     case pbnavitia::StreetNetworkMode::Car:
     case pbnavitia::StreetNetworkMode::Taxi:
-        build_mono_modal_journey(*journey, dp_request, params, pathedges, trip_leg, api, valhalla_service_url);
+        build_mono_modal_journey(*journey, dp_request, params, pathedges, trip_leg, api, valhalla_service_url, lez_list);
         break;
     case pbnavitia::StreetNetworkMode::Bss:
-        build_bss_journey(*journey, dp_request, params, pathedges, trip_leg, api, valhalla_service_url);
+        build_bss_journey(*journey, dp_request, params, pathedges, trip_leg, api, valhalla_service_url, lez_list);
```
[the remark this answers is not in the document]
Comment (kadhikari) [reply asgard/direct_path_response_builder.cpp]: :+1:

File: asgard/handler.cpp (hunk 3 above)
Comment (kadhikari) [reply asgard/handler.cpp]: :+1:

File: asgard/tests/handler_test.cpp
Comment (xlqian) [reply asgard/tests/handler_test.cpp:1]: done

File: asgard/tests/direct_path_response_builder_test.cpp
Comment (xlqian) [reply asgard/tests/direct_path_response_builder_test.cpp:1]: done

**Review by woshilapin — APPROVED** [review]

**Pull request merged**